python/lsst/pipe/base/pipelineIR.py

```diff
@@ -902,36 +969,97 @@ def to_file(self, filename: str):
         """
         self.write_to_uri(filename)
 
-    def write_to_uri(self, uri: Union[ButlerURI, str]):
+    def write_to_uri(
+        self,
+        uri: Union[ButlerURI, str],
+        *,
+        expanded_tasks: Optional[Mapping[str, TaskIR]] = None,
+    ):
         """Serialize this `PipelineIR` object into a yaml formatted string and
         write the output to a file at the specified uri.
 
         Parameters
         ----------
         uri: `str` or `ButlerURI`
             Location of document to write a `PipelineIR` object.
+        expanded_tasks : `Mapping` [ `str`, `TaskIR` ], optional
+            Mapping containing replacement `TaskIR` objects that capture the
+            fully-expanded configuration rather than a set of overrides, in
+            deterministic order.  When this is not `None`, the ``instrument``
+            and ``parameters`` sections are not written and all other sections
+            are sorted (using the order of the given tasks) to maximize the
+            extent to which equivalent pipelines will be written identically.
         """
-        butlerUri = ButlerURI(uri)
-        butlerUri.write(yaml.dump(self.to_primitives(), sort_keys=False).encode())
+        with ButlerURI(uri).open("w") as buffer:
+            yaml.dump(self.to_primitives(expanded_tasks=expanded_tasks), buffer, sort_keys=False)
 
-    def to_primitives(self) -> Dict[str, Any]:
-        """Convert to a representation used in yaml serialization"""
+    def to_primitives(self, expanded_tasks: Optional[Mapping[str, TaskIR]] = None) -> Dict[str, Any]:
+        """Convert to a representation used in yaml serialization
+
+        Parameters
+        ----------
+        expanded_tasks : `Mapping` [ `str`, `TaskIR` ], optional
```

Comment:
This bothered me, too, but we absolutely can't afford to apply instrument overrides again (since other overrides may have been applied on top of those that should not be overridden), and that's what would happen if we left the instrument directive in.  I think the alternative is to add a syntax for specifying an instrument while noting that its overrides have already been applied.  Would you prefer that?

Reply:
Yeah I would prefer some syntax for that, or even another top level key, like `instrument_name` that does nothing, but is persisted.

This comment got be thinking, because these are conceptually `finalized` pipelines I can see why we can not apply instruments a second time. However, these are not actually distinguishable from any other pipelines on load, there really is nothing that stops someone from doing additional overrides, such as passing something like --instrument on the command line, or even deriving a whole new pipeline from this as an import. I think this could be a useful feature (especially in testing for production when you need to quickly modify or something), but I also think it could be a foot gun if not taken into account. With this in mind, is there anything about this ticket you would like to re-evaluate?